contracts/staking/GoodFundManager.sol

```diff
@@ -189,103 +221,191 @@ contract GoodFundManager is DAOContract {
-     * Only the avatar can destroy the contract.
-     */
-  /**  function end() public onlyAvatar {
+		if (gdUBI > 0) {
+			//transfer ubi to avatar on sidechain via bridge
+			require(
+				token.transferAndCall(
+					bridgeContract,
+					gdUBI,
+					abi.encodePacked(ubiRecipient)
+				),
+				"ubi bridge transfer failed"
+			);
+		}
+		lastCollectedInterest = block.timestamp;
+		uint256 gasCostToMintReward = 200000; // Gas cost to mint GD reward to keeper hardcoded so should be changed according to calculations
+		uint256 totalUsedGas =
+			((initialGas - gasleft() + gasCostToMintReward) * 110) / 100; // We will return as reward 1.1x of used gas in GD
+		uint256 gdAmountToMint = getGasPriceInGD(totalUsedGas);
+		GoodReserveCDai(nameService.addresses(nameService.RESERVE()))
+			.mintRewardFromRR(
+			nameService.getAddress("CDAI"),
+			msg.sender,
+			gdAmountToMint
+		);
+		uint256 gasPriceIncDAI = getGasPriceInCDAI(initialGas - gasleft());
+		if (block.timestamp >= lastCollectedInterest + 5184000) {
+			// 5184000 is 2 months in seconds
+			require(
+				interest >= gasPriceIncDAI,
+				"Collected interest should be bigger than spent gas"
+			); // This require is necessary to keeper can not abuse this function
+		} else {
+			require(
+				interest >= 4 * gasPriceIncDAI,
+				"Collected interests should be at least 4 times bigger than gas cost since last call of this function sooner than 2 months"
+			);
+		}
+		emit FundsTransferred(
+			msg.sender,
+			nameService.addresses(nameService.RESERVE()),
+			_stakingContracts,
+			interest,
+			gdUBI,
+			gdAmountToMint
+		);
+	}
+
+	/**
+	 * @dev  Function that get addresses of staking contracts which interests
+	 * can be collected with the gas amount that provided as parameter
+	 * @param _maxGasAmount The maximum amount of the gas that keeper willing to spend collect interests
+	 * @return addresses of the staking contracts to the collect interests
+	 */
+	function calcSortedContracts(uint256 _maxGasAmount)
+		public
+		view
+		returns (address[] memory)
+	{
+		uint256 activeContractsLength = activeContracts.length;
+		address[] memory addresses = new address[](activeContractsLength);
+		uint256[] memory balances = new uint256[](activeContractsLength);
+		uint256 tempInterest;
+		uint256 totalInterest;
+		int256 i;
+		require(
+			activeContractsLength > 0,
+			"There should be at least one active staking contract"
+		);
+		for (i = 0; i < int256(activeContractsLength); i++) {
+			(tempInterest, , ) = StakingContract(activeContracts[uint256(i)])
+				.currentUBIInterest();
+			totalInterest += tempInterest;
+			if (tempInterest != 0) {
+				addresses[uint256(i)] = activeContracts[uint256(i)];
+				balances[uint256(i)] = tempInterest;
+			}
+		}
+		uint256 gasCostInCdai = getGasPriceInCDAI(_maxGasAmount); // Get gas price in cDAI so can compare with possible interest amount to get
+		require(
+			totalInterest >= gasCostInCdai,
+			"Current interest's worth less than spent gas worth"
+		);
```

Comment:
we dont need any requires in this view function just return empty address list if not passing conditions